### ISSUE TITLE
Runtime error needs Protelis context

### ISSUE THREAD

**Issue by jakebeal** [Protelis/Protelis#231]
This error is caused by passing a tuple into a hood operator without wrapping it with nbr.
It is not detected at compile-time, but at runtime (as it should be, given no type inference). The location of the error in the Protelis code, however, is not given:

```
[main] ERROR i.u.a.core.implementations.Engine - The simulation engine crashed.
org.protelis.lang.interpreter.util.ProtelisRuntimeException: org.protelis.lang.datatype.impl.ArrayTupleImpl cannot be cast to org.protelis.lang.datatype.Field
	at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:132)
	at org.protelis.lang.interpreter.impl.All.evaluate(All.java:63)
	at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:126)
	at org.protelis.vm.impl.SimpleProgramImpl.compute(SimpleProgramImpl.java:76)
	at org.protelis.vm.ProtelisVM.runCycle(ProtelisVM.java:45)
	at org.protelis.test.infrastructure.RunProtelisProgram.execute(RunProtelisProgram.java:80)
	at it.unibo.alchemist.model.implementations.reactions.AReaction.execute(AReaction.java:281)
	at it.unibo.alchemist.core.implementations.Engine.doStep(Engine.java:696)
	at it.unibo.alchemist.core.implementations.Engine.run(Engine.java:641)
	at org.protelis.test.InfrastructureTester.testSingleRun(InfrastructureTester.java:365)
	at org.protelis.test.InfrastructureTester.generalTest(InfrastructureTester.java:218)
	at org.protelis.test.InfrastructureTester.runTest(InfrastructureTester.java:296)
	at org.protelis.lang.test.TestApis.test(TestApis.java:24)
	at org.protelis.lang.test.TestApis.test(TestApis.java:20)
	at org.protelis.lang.test.TestApis.testUnionHoodPlusSelf(TestApis.java:1097)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:305)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:365)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:330)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:78)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:328)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:65)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:292)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:305)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:412)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
Caused by: java.lang.ClassCastException: org.protelis.lang.datatype.impl.ArrayTupleImpl cannot be cast to org.protelis.lang.datatype.Field
	at org.protelis.lang.interpreter.impl.HoodCall.evaluate(HoodCall.java:57)
	at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:126)
	... 42 common frames omitted
```

**Comment by DanySK** [comment]
I tried to reproduce this one with a test:
```java
    /**
     * Test issue #231.
     */
    @Test
    public void testRuntimeErrorOnClassCastFailure() {
        ProgramTester.runExpectingErrors("minHood([])", ProtelisRuntimeException.class, e -> {
            assertNotNull(e.getMessage());
            final String fullTrace = e.toString();
            assertTrue("Exception does not include type cast failure indication\n" + fullTrace,
                    e.getMessage().contains("cannot be cast"));
        });
    }
```
but it works, and reports the Protelis stack trace as expected. Can you provide some more contextual information? The first line seems to indicate the error to be happening inside (a very old version of) Alchemist. Are you adding tests in protelis-lang? We may need to upgrade out test infrastructure dependencies.

**Comment by jakebeal** [comment]
Here's a simple test that fails on my machine.

In protelis-lang, change `src/test/resources/boundBroadast.pt` to be:

````
import protelis:coord:spreading
boundBroadcast(!env.has("obstacle"), "bar", env.has("source"), "foo");

minHood("foo")
````

Obviously, the minHood("foo") should fail.  Its error when I run as JUnit is: 
````
[main] ERROR i.u.a.core.implementations.Engine - The simulation engine crashed.
org.protelis.lang.interpreter.util.ProtelisRuntimeException: java.lang.String cannot be cast to org.protelis.lang.datatype.Field
	at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:129)
	at org.protelis.lang.interpreter.impl.All.lambda$0(All.java:52)
	at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.forEachWithIndex(AbstractAnnotatedTree.java:168)
	at org.protelis.lang.interpreter.impl.All.evaluate(All.java:50)
	at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123)
	at org.protelis.vm.impl.SimpleProgramImpl.compute(SimpleProgramImpl.java:75)
	at org.protelis.vm.ProtelisVM.runCycle(ProtelisVM.java:45)
	at org.protelis.test.infrastructure.RunProtelisProgram.execute(RunProtelisProgram.java:80)
	at it.unibo.alchemist.model.implementations.reactions.AReaction.execute(AReaction.java:281)
	at it.unibo.alchemist.core.implementations.Engine.doStep(Engine.java:696)
	at it.unibo.alchemist.core.implementations.Engine.run(Engine.java:641)
	at org.protelis.test.InfrastructureTester.testSingleRun(InfrastructureTester.java:365)
	at org.protelis.test.InfrastructureTester.generalTest(InfrastructureTester.java:218)
	at org.protelis.test.InfrastructureTester.runTest(InfrastructureTester.java:296)
	at org.protelis.lang.test.TestApis.test(TestApis.java:24)
	at org.protelis.lang.test.TestApis.test(TestApis.java:20)
	at org.protelis.lang.test.TestApis.testBoundBroadcast(TestApis.java:147)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:305)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:365)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:330)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:78)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:328)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:65)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:292)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:305)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:412)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
Caused by: java.lang.ClassCastException: java.lang.String cannot be cast to org.protelis.lang.datatype.Field
	at org.protelis.lang.interpreter.impl.HoodCall.evaluate(HoodCall.java:56)
	at org.protelis.lang.interpreter.impl.AbstractAnnotatedTree.eval(AbstractAnnotatedTree.java:123)
	... 44 common frames omitted
````

**Comment by jakebeal** [comment]
Should be addressed by #270

**Comment by DanySK** [comment]
Fixed in [13.1.0](https://github.com/Protelis/Protelis/releases/tag/13.1.0)
